### PR TITLE
Command line interface with login and permission control

### LINKED ISSUE #1: Log in/log out system (username, password)
Log in/log out system (username, password)
Basic CLI
Log in/out via command line
Permission control

### LINKED ISSUE #2: Permission control

### PR DESCRIPTION
**Pull request by dennisylyung**
Main command line interface using the [Cmd library](https://docs.python.org/3/library/cmd.html) from core python. 
The interface definition is placed in the `interfaces` directory, with the main loop called in `\__main__.py`. 
To run the CLI, run `__main__.py` at the project directory.

User login is implemented in `models\user.py`, as a base class for role types. 
Password is stored as a salted hash value. 

Permission control is implemented as a python decorator `@require_permission`, located in `models\user.py`. 
As examples, commands `profile` and `create_volunteer` are implemented, the former open to all roles, while the later requires the role `Admin`. 
Hence, role classes `Admin` and `Volunteer` are also created. This may cause a merge conflict with #13 but could be resolved easily. 

Special handling was done to support "two-word commands" for the CLI, like `create volunteer`, `create plan`, or `list plans`. 
To implement two-word commands, make sure the first word is included in `EmsShell.action_words`, then implement a command method accordingly. 

Demo:
image

Closes #1, closes #2

### CODE REVIEW

File: interfaces/cli.py
Comment (Elynor-Kamil) [inline interfaces/cli.py:42]: Remove the :remove: please

**Review by Elynor-Kamil — APPROVED** [review]

**Review by vwcchan — APPROVED** [review]

**Review by Elynor-Kamil — APPROVED** [review]

**Review by Shalaka-Yerawadekar — APPROVED** [review]

**Review by YunsyYin — APPROVED** [review]
here we also have volunteer.py. Will it conflict with the file in the volonteerInfo branch after the merge?

**Comment by dennisylyung** [comment]
> here we also have volunteer.py. Will it conflict with the file in the volonteerInfo branch after the merge?

Actually yes. 
I mainly put the Volunteer class there as a demo of how the CLI work. 
I am expecting to merge this after the volunteer PR, and manually resolve the merge conflict then.

**Review by Shalaka-Yerawadekar — APPROVED** [review]

**Review by michellechan1123 — APPROVED** [review]

**Review by YingboZhou — APPROVED** [review]

**Comment by dennisylyung** [comment]
The demo command `create user` has been removed since it is incompatible with the actual `Volunteer` class. 
The volunteer creation command should be implemented later on. 
(cc: @YunsyYin)

Demo code:
```python
    @require_role(Admin)
    def do_create_volunteer(self, arg):
        """
        Create a new volunteer user
        """
        username = input('Username: ')
        password = input('Password: ')
        user = Volunteer(username, password)
        if user in users_catalog.values():
            print('Username already exists')
        else:
            users_catalog[username] = user  # TODO: persist new user
            print(f'User {username} created')
```

**Pull request merged**
